### PR TITLE
[GStreamer] Add support for DMA-BUF to GL video sink

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### 36eefe856208603fc8b66a8e111defb51407de48
<pre>
[GStreamer] Add support for DMA-BUF to GL video sink
<a href="https://bugs.webkit.org/show_bug.cgi?id=279672">https://bugs.webkit.org/show_bug.cgi?id=279672</a>

Reviewed by Alejandro G. Castro and Xabier Rodriguez-Calvar.

By adding DMA-BUF support to GL video sink, we can fallback to GL memory
when gst can&apos;t provide DMA-BUF memory, instead of mapping the frame to
upload it as a DMA-BUF buffer. With the new platform layer buffers we
don&apos;t need specific TextureMapperPlatformLayerProxy for DMA-BUF, we can
add support for YUV formats to CoordinatedPlatformLayerBufferDMABuf and
use the existing GL proxy. This simplifies the code and we can remove a
lot of code that is now unsued.

* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::dmabufFormatsForVideo):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
(WebCore::queryDMABufFormats):
(WebCore::GLDisplay::dmabufFormats):
(WebCore::GLDisplay::dmabufFormatsForVideo):
* Source/WebCore/platform/graphics/egl/GLDisplay.h:
* Source/WebCore/platform/graphics/gbm/DMABufBuffer.h:
* Source/WebCore/platform/graphics/gbm/DMABufColorSpace.h: Removed.
* Source/WebCore/platform/graphics/gbm/DMABufEGLUtilities.h: Removed.
* Source/WebCore/platform/graphics/gbm/DMABufFormat.h: Removed.
* Source/WebCore/platform/graphics/gbm/DMABufObject.h: Removed.
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h: Removed.
* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp: Removed.
* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/DMABufUtilities.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/DMABufUtilities.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/DMABufVideoSinkGStreamer.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/DMABufVideoSinkGStreamer.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp:
(initializeDMABufAvailability):
(buildDMABufCaps):
(webKitGLVideoSinkConstructed):
(webKitGLVideoSinkProbePlatform):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):
(WebCore::MediaPlayerPrivateGStreamer::pushDMABufToCompositor): Deleted.
(WebCore::MediaPlayerPrivateGStreamer::createVideoSinkDMABuf): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
(WebCore::VideoFrameGStreamer::convert):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.h: Removed.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.cpp:
(WebCore::importToTexture):
(WebCore::formatIsYUV):
(WebCore::yuvFormatPlaneInfo):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::importYUV const):
(WebCore::CoordinatedPlatformLayerBufferDMABuf::importDMABuf const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferDMABuf.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
(WebCore::CoordinatedPlatformLayerBufferVideo::create):
(WebCore::CoordinatedPlatformLayerBufferVideo::CoordinatedPlatformLayerBufferVideo):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferIfNeeded):
(WebCore::videoFormatToDRMFourcc):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromDMABufMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::createBufferFromGLMemory):
(WebCore::CoordinatedPlatformLayerBufferVideo::paintToTextureMapper):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.cpp:
(WebCore::CoordinatedPlatformLayerBufferYUV::create):
(WebCore::CoordinatedPlatformLayerBufferYUV::CoordinatedPlatformLayerBufferYUV):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferYUV.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/283836@main">https://commits.webkit.org/283836@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5949b4e6017d9f920860ad22e44edb021a078f5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67559 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71607 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18696 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54736 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18500 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12502 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43076 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58425 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/67071 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15830 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17054 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/60674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/61707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73306 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66804 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11518 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/15486 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61549 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11553 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58494 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9405 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3018 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88573 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10261 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15629 "Found 8 new JSC stress test failures: wasm.yaml/wasm/fuzz/memory.js.wasm-no-cjit, wasm.yaml/wasm/stress/js-to-wasm-i64-stack.js.wasm-eager, wasm.yaml/wasm/stress/js-to-wasm-many-i64.js.wasm-bbq, wasm.yaml/wasm/stress/js-to-wasm-many-i64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/js-wasm-call-many-return-types-on-stack-no-args.js.default-wasm, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-slow-memory, wasm.yaml/wasm/stress/tail-call-js-inline.js.wasm-eager-jettison, wasm.yaml/wasm/stress/tail-call.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43820 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43561 "Built successfully") | | |

### CODE REVIEW

**Comment by philn** [comment]
IIUC one nice thing with the DMABufObject abstraction is that it could be used for the GPUProcess, whenever we move GStreamer there.

How is that going to work now?

**Comment by philn** [comment]
If DMABufVideoSinkGStreamer.cpp is removed please also remove its references from Tools/Scripts/webkitpy/style/checker.py

**Comment by philn** [comment]
Also please remove the WEBKIT_GST_DMABUF_SINK_DISABLED handling from Tools/Scripts/webkitpy/port/glib.py

**Comment by philn** [comment]
Maybe I got confused between DMABufObject and DMABufBuffer :)

**Comment by philn** [comment]
This is likely to solve negotiation with nvidia decoders, i think... the fallback to GL from decoder to sink should work there now.

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp [hunk 1]
```diff
@@ -26,8 +26,14 @@
 #include "GStreamerVideoSinkCommon.h"
 #include "PlatformDisplay.h"
 #include <gst/gl/gl.h>
+#include <wtf/glib/GUniquePtr.h>
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp]: You don't need this, it's brought in cascade from GStreamerCommon.h

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp [hunk 2]
```diff
@@ -57,6 +63,88 @@ static GstStaticPadTemplate sinkTemplate = GST_STATIC_PAD_TEMPLATE("sink", GST_P
...
+
+static void initializeDMABufAvailability()
+{
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        if (!webkitGstCheckVersion(1, 20, 0))
+            return;
+
+        const char* value = g_getenv("WEBKIT_GST_DMABUF_SINK_DISABLED");
+        auto valueSpan = WTF::span(value);
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp]: You can collapse these two lines

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp [hunk 3]
```diff
@@ -57,6 +63,88 @@ static GstStaticPadTemplate sinkTemplate = GST_STATIC_PAD_TEMPLATE("sink", GST_P
...
+    });
+}
+
+static GRefPtr<GstCaps> buildDMABufCaps()
+{
+    GRefPtr<GstCaps> caps = adoptGRef(gst_caps_from_string("video/x-raw(memory:DMABuf), width = " GST_VIDEO_SIZE_RANGE ", height = " GST_VIDEO_SIZE_RANGE ", framerate = " GST_VIDEO_FPS_RANGE));
+#if GST_CHECK_VERSION(1, 24, 0)
+    gst_caps_set_simple(caps.get(), "format", G_TYPE_STRING, "DMA_DRM", nullptr);
+
+    static const char* formats = getenv("WEBKIT_GST_DMABUF_FORMATS");
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp]: Use g_getenv

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp [hunk 4]
```diff
@@ -57,6 +63,88 @@ static GstStaticPadTemplate sinkTemplate = GST_STATIC_PAD_TEMPLATE("sink", GST_P
...
+    if (formats && *formats) {
+        GUniquePtr<char*> tokens(g_strsplit(formats, ",", -1));
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp:90]: You could use StringView and split.

**Review by calvaris — APPROVED** [review]
I reviewed the GStreamer part, somebody has to review the rest.

**Comment by carlosgcampos** [comment]
> IIUC one nice thing with the DMABufObject abstraction is that it could be used for the GPUProcess, whenever we move GStreamer there.
> 
> How is that going to work now?

There's already simpler code in main to handle that, used to implement WebGL.

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp (hunk 3 above)
Comment (carlosgcampos) [reply Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp]: why?

File: Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp (hunk 4 above)
Comment (carlosgcampos) [reply Source/WebCore/platform/graphics/gstreamer/GLVideoSinkGStreamer.cpp:90]: I wanted to avoid utf8 -> utf16 -> utf8 conversions.

**Comment by philn** [comment]
> > IIUC one nice thing with the DMABufObject abstraction is that it could be used for the GPUProcess, whenever we move GStreamer there.
> > How is that going to work now?
> 
> There's already simpler code in main to handle that, used to implement WebGL.

Can you update `Tools/Scripts/generate-gpup-webgl` then?

**Comment by carlosgcampos** [comment]
Sure, I don't think we are autogenerating that part, though.

**Comment by carlosgcampos** [comment]
https://github.com/WebKit/WebKit/pull/33694

**Comment by septatrix** [comment]
I cannot get this branch to build. It is stuck for several hours on

```txt
[8133/8357][8848.742] : && /usr/bin/clang++ -fdiagnostics-color=always -fcolor-diagnostics -Wextra -Wall -pipe -Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-parentheses-equality -Qunused-arguments -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fdebug-types-section -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections -DNDEBUG -fuse-ld=mold -Wl,--gc-sections -Wl,--disable-new-dtags @CMakeFiles/TestWebCore.rsp -o bin/TestWebKitAPI/TestWebCore  && :
```

See https://copr.fedorainfracloud.org/coprs/septatrix/wpewebkit/build/8021871/

**Comment by septatrix** [comment]
> COPR is often overloaded, I doubt this timeout is related with this PR.

I would understand that if it times out in general but that compilation (and the previous one) is stuck on that exact step since 8848.742s into invoking make (i.e. 2.45798h). And it is now running for 17h so it has been stuck on that single step for about 15h (and the previous build with smaller timeout threshold got stuck at that very same command).

**Comment by philn** [comment]
Did you check the log file?

The build completes just fine, the issue seems to be related with your RPM:
```
RPM build errors:
    File not found: /builddir/build/BUILDROOT/wpewebkit-2.45.6-1.fc40.x86_64/usr/share/doc/wpewebkit/NEWS
```

How did you make that tarball?

**Comment by septatrix** [comment]
> Did you check the log file?
> 
> The build completes just fine, the issue seems to be related with your RPM:
> 
> ```
> RPM build errors:
>     File not found: /builddir/build/BUILDROOT/wpewebkit-2.45.6-1.fc40.x86_64/usr/share/doc/wpewebkit/NEWS
> ```

It seems like I checked to log file to quick before it really terminated and only saw it being stuck at that one line which I now think is likely the linking step or something else which is expected to take a while.

> How did you make that tarball?

I dowloaded the tar from GitHub though these do not seem to include a NEWS file. I now adjusted the rpm spec. There seem to be a few other (packaging) errors now but the build itself seems to run fine. Sorry for the confusion

**Review by alexgcastro — APPROVED** [review]
LGTM! Great clean-up thanks to use the new buffer architecture.

**Pull request merged**